src/runtime/node/internal/async_hooks/async-local-storage.ts

```diff
@@ -30,21 +30,42 @@ class _AsyncLocalStorage<T> implements nodeAsyncHooks.AsyncLocalStorage<T> {
     callback: (...args: TArgs) => R,
     ...args: TArgs
   ): R {
+    const previousStore = this._currentStore;
     this._currentStore = store;
-    const res = callback(...args);
-    this._currentStore = undefined;
-    return res;
+    try {
+      const res = callback(...args);
+      if (res != null && typeof (res as any).then === "function") {
+        return Promise.resolve(res).finally(() => {
+          this._currentStore = previousStore;
```

Comment:
This is dangerous when two promises resolve in parallel, as there is no concurrency lock.